client/client_test.go

```diff
@@ -61,3 +66,84 @@ func TestSubscribeHeaders(t *testing.T) {
 	_, ok := <-headers
 	assert.Equal(t, false, ok)
 }
+
+func TestSubmitTx(t *testing.T) {
+	c := NewClient(NewLocalConnection(node))
+	ctx := context.Background()
+
+	key := cmn.RandStr(10)
+	tx := &KvTx{Key: key}
+	mem, err := c.SubmitTx(ctx, tx)
+	assert.Nil(t, err)
+	assert.Nil(t, mem.Err)
```

Comment:
This is what I had in mind about `SubmitTx`. You have to check two places for an error, although they are exclusive (only one can be set).

Reply:
Yeah, I'm not sure how to handle this better.

I have tried making ErrNetwork, but I am not really happy with that either (it is any error the server returns as well... like invalid request).

mem.Err is when it got the packet and the abci app rejected it
err is when we couldn't even post - network error, incompatible tendermint version, etc

I could combine them (either all in mem.Err or both in err). In either case, it didn't make it into mempool.

For the CommitTx, there is a large difference. `(res, err)`.
`err` means it never made it into the block - network error, rejected in submit, etc
`res.Err` means it was in the block, but failed during DeliverTx (updating nonce, deducting fees, etc).

I guess I could remove mem.Err and make any case that did not result in mempool as err. But leave the double-check for CommitTx (res.Err means got in block but failed, err means never made it to the block)

what do you think?

Reply:
I think combining them is fine. As a user of this client do you really care why the operation failed? You may want to know if you should retry or not although I expect that the usual handling of this error will be to log it and quit the program.

In the case of the `CommitTx`, it is returning multiple responses and they can succeed or fail. There it is mandatory for every response to contain a separate error.